### DISCUSSION TITLE
How does PyMuPDF differ bookmark with outline?

### DISCUSSION THREAD

**Discussion by tristone13th** [pymupdf/PyMuPDF#1714] [Q&A]
In the official documentation, bookmark and outline are the same thing, but some apps may implement the so called bookmark in their way, how does PyMuPDF retrieve these information?

for reference, please see:
[Understanding the Difference between Bookmarks and the Outline in a PDF \| PSPDFKit](https://pspdfkit.com/blog/2019/understanding-pdf-outline/)

**Comment by JorjMcKie** [comment]
Obviously, (Py-) MuPDF cares about the official PDF specification only (which no longer is copyright of Adobe but of the official standard ISO 32000-1!). And rightly so - how could it be otherwise? It takes no note of specialized behaviors of whatever PDF viewer - including Adobe's Acrobat, which is not part of any standard of an official body.

While it  may be a careless use of language identifying the two words "outline" and "bookmark", a few comments may add clarification:
* The items in the PDF's outline tree are technically **_links_** that are organized in a special (overly complex) internal structure.
* Just like any link, the may or **_may not_** point to locations inside the current document. But they may also point to other documents or be URLs (which then exceeds the ususal perception of a "bookmark").
* Everywhere else in a PDF, there may independently exist links which do quite the same: point to other pages inside the document or be more general URLs. With all justification, **_internal_** links **_can_** be called bookmarks - and are fully supported.
* Using PyMuPDF, you can modify a PDF's outline tree as a whole (by entirely replacing it with a new version), or change single items. In any case, the result is a standards conforming outline tree that should be supported by all PDF viewers. The same is true for link annotations.

Don't forget: MuPDF does not only support PDFs, but several other document types, too.
Where present at all, the specific notions of outline trees and links for all of these are supported by MuPDF as well - with the same unified API.

> On a side note:
MuPDF also has its own notion of bookmark when it comes to document types with variable layout, like e-books: You can compute a page's physical location in the document (then called a "bookmark"), re-layout the document, then find the new page number of that previously computed bookmark.
This is supported by PyMuPDF as well.

If you take the position of a book reading individual, the notion _bookmark_ may take on the meaning of a marker of your personal reading progress.
This is something your viewer may support in whatever specific way: stored outside the document or via links inside. Only the latter case is supported by MuPDF.

**Reply by tristone13th** [reply]
Thanks your amazing answer. I think the bookmarks are not stored outside for when I send it to another computer, it's still there. But I cannot find the link. Specially, I'm using Xodo.
